### PR TITLE
feat: add `FieldMerkleTree`

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [FieldMerkleTree](https://github.com/Plonky3/Plonky3/blob/main/merkle-tree/src/merkle_tree.rs).

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole Please fix typos in 632070f. arrayo_to_vector -> array_to_vector, contaier_util -> container_util.

File: tachyon/math/matrix/BUILD.bazel [hunk 1]
```diff
@@ -48,5 +52,14 @@ tachyon_cc_unittest(
...
+    ] + select({
+        "@platforms//cpu:x86_64": [
+            "//tachyon/math/finite_fields/baby_bear:packed_baby_bear_avx2",
+            "//tachyon/math/finite_fields/baby_bear:packed_baby_bear_avx512",
+        ],
+        "@platforms//cpu:aarch64": [
+            "//tachyon/math/finite_fields/baby_bear:packed_baby_bear_neon",
+        ],
+        "//conditions:default": [],
+    }),
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/matrix/BUILD.bazel:65]: The reason why I do like this is because I wanted to test `avx2` and `avx512` at the same time on x86 if it is built with `avx512`. But I think in this context, just using `PackedBabyBear` seems like to make more sense.

File: tachyon/base/containers/container_util.h [hunk 2]
```diff
@@ -288,6 +288,36 @@ std::vector<std::vector<T>> Array2DToVector2D(const T (&arr)[N][M]) {
...
 }
 
+template <size_t N, typename Generator,
+          typename FunctorTraits = internal::MakeFunctorTraits<Generator>,
+          typename RunType = typename FunctorTraits::RunType,
+          typename ReturnType = typename FunctorTraits::ReturnType,
+          typename ArgList = internal::ExtractArgs<RunType>,
+          size_t ArgNum = internal::GetSize<ArgList>,
+          std::enable_if_t<ArgNum == 0>* = nullptr>
+std::array<ReturnType, N> CreateArray(Generator&& generator) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/base/containers/container_util.h:298]: Like in `CreateVector()`, in that case, I think you need to use `base::Parallelize()` or `OPENMP_PARALLEL_FOR`

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — CHANGES_REQUESTED** [review]
https://github.com/kroma-network/tachyon/pull/436/commits/6154023bf3ab1972aa523be130aacd59af8bef61
Just for clarity, could you add as a commit message “enables Matrix<F>::Random()” or something similar?

https://github.com/kroma-network/tachyon/pull/436/commits/7b75af46576f70f474d33d7c4da2ef316340f2f0
I think you missed `sparse_matrix_unittest.cc`

File: tachyon/math/matrix/matrix_utils_unittest.cc [hunk 3]
```diff
@@ -21,4 +23,71 @@ TEST_F(MatrixUtilsTest, Circulant) {
   EXPECT_EQ(circulant, expected);
 }
 
+class MatrixPackingTest : public FiniteFieldTest<PackedBabyBear> {};
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/matrix/matrix_utils_unittest.cc:25]: I can run this with TypedTest, but I don't think this approach is meaningful. I'm concerned that it will just increase the test running time, and the logic we want to test here doesn't depend on a specific packed prime field implementation. Even though more packed prime fields are introduced, we're unlikely to add them to this test.

File: tachyon/math/matrix/matrix_utils_unittest.cc [hunk 4]
```diff
@@ -21,4 +23,71 @@ TEST_F(MatrixUtilsTest, Circulant) {
...
+    ASSERT_EQ(packed_values.size(), 2);
+    for (size_t i = 0; i < packed_values.size(); ++i) {
+      for (size_t j = 0; j < N; ++j) {
+        EXPECT_EQ(packed_values[i][j], matrix(R, i * N + j));
+      }
+    }
+  }
+
+  {
+    math::Matrix<BabyBear, 2 * N - 1, 2 * N - 1> matrix;
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/matrix/matrix_utils_unittest.cc]: We can't do like above, since the types of `matrix` are different.

[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/matrix/matrix_utils_unittest.cc]: Oh, yes if it is dynamic, it's possible. I'll apply!

**Comment by chokobole** [comment]
> [6154023](https://github.com/kroma-network/tachyon/commit/6154023bf3ab1972aa523be130aacd59af8bef61) Just for clarity, could you add as a commit message “enables Matrix::Random()” or something similar?
> 
> [7b75af4](https://github.com/kroma-network/tachyon/commit/7b75af46576f70f474d33d7c4da2ef316340f2f0) I think you missed `sparse_matrix_unittest.cc`

I explictly don't apply the change to `sparse_matrix_unittest.cc` since it doesn't include `matrix_types.h`.

File: tachyon/math/matrix/matrix_utils_unittest.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/matrix/matrix_utils_unittest.cc]: I changed but this includes `remaining_values.clear()` before the second call to `PackRowHorizontally`

File: tachyon/crypto/hashes/sponge/poseidon/poseidon.h [hunk 5]
```diff
@@ -12,7 +12,6 @@
 #include "tachyon/base/buffer/copyable.h"
 #include "tachyon/crypto/hashes/sponge/poseidon/poseidon_config.h"
 #include "tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h"
-#include "tachyon/crypto/hashes/sponge/sponge_state.h"
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon.h:15]: Why was this removed? Doesn't `ApplyARK` and `ApplyMix` still use these?

File: tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h [hunk 6]
```diff
@@ -15,7 +15,6 @@
 #include "tachyon/crypto/hashes/sponge/poseidon2/poseidon2_config.h"
 #include "tachyon/crypto/hashes/sponge/poseidon2/poseidon2_horizen_internal_matrix.h"
 #include "tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h"
-#include "tachyon/crypto/hashes/sponge/sponge_state.h"
```
Comment (ashjeong) [inline tachyon/crypto/hashes/sponge/poseidon2/poseidon2.h:18]: Why was this removed? Doesn't `ApplyARK` and `ApplyMix` still use these?

File: tachyon/zk/plonk/examples/BUILD.bazel [hunk 7]
```diff
@@ -57,7 +57,6 @@ tachyon_cc_library(
         ":simple_circuit_test_data",
         ":simple_lookup_circuit",
         ":simple_lookup_circuit_test_data",
```
Comment (ashjeong) [inline tachyon/zk/plonk/examples/BUILD.bazel:59]: Could you include `container_util.h` here to be explicit?

Comment (chokobole) [reply tachyon/zk/plonk/examples/BUILD.bazel:59]: But `circuit_test.h` doesn't use any thing from `container_util.h`, Rather, I found that `circuit_test` target depends on the unused dependencies.

**Review by ashjeong — APPROVED** [review]
LGTM
